cachi2/core/package_managers/yarn_classic/main.py

```diff
@@ -125,3 +134,18 @@ def _verify_corepack_yarn_version(source_dir: RootedPath, env: dict[str, str]) -
         )
 
     log.info("Processing the request using yarn@%s", installed_yarn_version)
+
+
+def _verify_no_offline_mirror_collisions(packages: Iterable[YarnClassicPackage]) -> None:
+    """Verify that there are no duplicate tarballs in the offline mirror."""
+    tarballs = []
+    for p in packages:
+        if isinstance(p, UrlPackage):
```

Comment:
Do we want to handle collisions between other types of Packages as well? For example, I could have a dependency from the registry that results in cached tarball named foo-1.0.0.tgz and have a url dependency with the same name in the path

Reply:
I added registry packages to check as well. It should be all I believe or?

Reply:
IIRC git packages are the only additional package type that is copied into the offline mirror. The others (file, workspace, link) are just present in the source repo.

Does this check also work for scoped packages? For example, in the main integration test, `@colors/colors` is one of the dependencies with a resolved url `https://registry.yarnpkg.com/@colors/colors/-/colors-1.6.0.tgz#ec6cd237440700bc23ca23087f513c75508958b0` and a tarball in the offline mirror `@colors-colors-1.6.0.tgz`

We may need to incorporate some of the logic found here: https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/fetchers/tarball-fetcher.js#L77

Reply:
I tried to rewrite the implementation in yarnpkg to Python.

> IIRC git packages are the only additional package type that is copied into the offline mirror. The others (file, workspace, link) are just present in the source repo.

Added git packages too, even though it seems an extremely rare scenario because of hash being added to the name of the repo.

Reply:
Nice job, thanks for taking care of this :100: